termwiz/src/render/terminfo.rs

```diff
@@ -52,32 +52,37 @@ impl TerminfoRenderer {
         allow(clippy::cyclomatic_complexity, clippy::cognitive_complexity)
     )]
     fn flush_pending_attr<W: UnixTty + Write>(&mut self, out: &mut W) -> Fallible<()> {
-        macro_rules! attr_on_off {
-            ($cap_on:ident, $cap_off:ident, $attr:expr, $accesor:ident, $sgr:ident) => {
-                let value = $attr.$accesor();
-                if value != self.current_attr.$accesor() {
-                    let out = out.by_ref();
-                    let on: bool = value.into();
-                    if on {
-                        if let Some(attr) = self.get_capability::<cap::$cap_on>() {
-                            attr.expand().to(out)?;
-                        } else {
-                            write!(out, "{}", CSI::Sgr(Sgr::$sgr(value)))?;
-                        }
-                    } else {
-                        if let Some(attr) = self.get_capability::<cap::$cap_off>() {
-                            attr.expand().to(out)?;
-                        } else {
-                            write!(out, "{}", CSI::Sgr(Sgr::$sgr(value)))?;
-                        }
-                    }
+        macro_rules! attr_on {
+            ($cap:ident, $sgr:expr) => {
+                if let Some(attr) = self.get_capability::<cap::$cap>() {
+                    attr.expand().to(out.by_ref())?;
+                } else {
+                    write!(out, "{}", CSI::Sgr($sgr))?;
                 }
             };
+            ($sgr:expr) => {
+                write!(out, "{}", CSI::Sgr($sgr))?;
+            };
         }
 
         if let Some(attr) = self.pending_attr.take() {
+            let mut current_foreground = self.current_attr.foreground;
+            let mut current_background = self.current_attr.background;
+
             if !attr.attribute_bits_equal(&self.current_attr) {
-                if let Some(sgr) = self.get_capability::<cap::SetAttributes>() {
+                // Updating the attribute bits also resets the colors.
+                current_foreground = ColorAttribute::Default;
+                current_background = ColorAttribute::Default;
+
+                // The SetAttributes capability can only handle single underline and slow blink.
+                let sgr_cap =
+                    if attr.underline() != Underline::Double && attr.blink() != Blink::Rapid {
+                        self.get_capability::<cap::SetAttributes>()
+                    } else {
+                        None
+                    };
+
+                if let Some(sgr) = sgr_cap {
                     sgr.expand()
                         .bold(attr.intensity() == Intensity::Bold)
                         .dim(attr.intensity() == Intensity::Half)
```

Comment:
Now that I look at this, the old code was treating double underline as equivalent to single here, and any blink as blink.  I'm ok with that as a first approximation, and then we can output the explicit double-underline and fast-blink as followups if needed.

Reply:
Hmm.. we could drop them by testing for `attr.underline() == Underline::Single` (so we only set underline if it is single underline), and then later on do `if attr.underline() == Underline::Double { attr_on!(Sgr::Underline(Underline::Double))` (like we do for italics).  How does that sound?